askama_parser/src/lib.rs

```diff
@@ -350,13 +350,112 @@ fn str_lit(i: &str) -> ParseResult<'_> {
     Ok((i, s.unwrap_or_default()))
 }
 
+/// If this is an ascii or unicode endpoint, it has a maximum value which is checked in this
+/// function.
+///
+/// If it fails, we return `Err` which contains the error message.
+fn validate_endpoint(
+    s: &str,
+    prefix: &str,
+    max_value: u32,
+    err1: &'static str,
+    err2: &'static str,
+) -> Result<(), &'static str> {
+    let Some(mut nb) = s.strip_prefix(prefix) else {
+        return Ok(());
+    };
+    if prefix.ends_with('{') {
+        if let Some(without_bracket) = nb.strip_suffix('}') {
+            nb = without_bracket;
+        }
+    }
+    let Ok(nb) = u32::from_str_radix(nb, 16) else {
+        return Err(err1);
+    };
+    if nb > max_value {
+        return Err(err2);
+    }
+    Ok(())
+}
+
+// Information about allowed character escapes is available at:
+// <https://doc.rust-lang.org/reference/tokens.html#ascii-escapes>.
```

Comment:
I think we should reference https://doc.rust-lang.org/reference/tokens.html#character-literals here. It feels like `recognize()` should yield something more like an `enum Char { CharLiteral(_), QuoteEscape(_), AsciiEscape(_), UnicodeEscape(_) }`, then we can add inline code that applies extra validation for the latter two variants.

Right now this code just passes different lists of 4 arguments that actually all depend on the single bit of ASCII vs Unicode so I think we should make that more clear.